src/main/java/com/zaxxer/hikari/pool/HikariPool.java

```diff
@@ -631,13 +631,24 @@ public void run()
          }
 
          logPoolState("Before cleanup ");
-         for (PoolBagEntry bagEntry : connectionBag.values(STATE_NOT_IN_USE)) {
+         boolean evicted = false;
+         List<PoolBagEntry> bag = connectionBag.values(STATE_NOT_IN_USE);
+         for (PoolBagEntry bagEntry : bag) {
+            if (bagEntry.evicted) {
+               closeConnection(bagEntry, "(connection evicted)");
+               evicted = true;
+            }
+         }
+         if (evicted) {
+        	//get fresh bag
+            bag = connectionBag.values(STATE_NOT_IN_USE);
+         }
+         int removable = bag.size() - config.getMinimumIdle();
+         for (PoolBagEntry bagEntry : bag) {
             if (connectionBag.reserve(bagEntry)) {
-               if (bagEntry.evicted) {
-                  closeConnection(bagEntry, "(connection evicted)");
-               }
-               else if (idleTimeout > 0L && clockSource.elapsedMillis(bagEntry.lastAccess, now) > idleTimeout) {
+               if (removable > 0 && idleTimeout > 0L && clockSource.elapsedMillis(bagEntry.lastAccess, now) > idleTimeout) {
                   closeConnection(bagEntry, "(connection passed idleTimeout)");
+                  removable--;
```

Comment:
That's not going to work, and is overly complex.  It won't work because ``closeConnection()`` is *asynchronous*, so even if evicted connections are "closed" the second call to get STATE_NOT_IN_USE bag entries is going to return the exact same collection 99.9% of the time.  Take my suggested code, and simply remove the decrement of ``removable`` from the evicted path, its better and simpler:
```java
final List<PoolBagEntry> bag = connectionBag.values(STATE_NOT_IN_USE);
int removable = bag.size() - config.getMinimumIdle();
for (PoolBagEntry bagEntry : bag) {
   if (connectionBag.reserve(bagEntry)) {
      if (bagEntry.evicted) {
         closeConnection(bagEntry, "(connection evicted)");
      }
      else if (removable > 0 && idleTimeout > 0L && clockSource.elapsedMillis(bagEntry.lastAccess, now) > idleTimeout){
         closeConnection(bagEntry, "(connection passed idleTimeout)");
         removable--;
      }
      else {
         connectionBag.unreserve(bagEntry);
      }
   }
}
```